tests/src/junit/java/net/neoforged/neoforge/unittest/transfer/InstancedHandlerTests.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.unittest.transfer;
+
+import net.minecraft.world.level.block.Blocks;
+import net.minecraft.world.level.material.Fluids;
+import net.neoforged.neoforge.transfer.ResourceHandlerUtil;
+import net.neoforged.neoforge.transfer.handlers.templates.EmptyResourceHandler;
+import net.neoforged.neoforge.transfer.handlers.templates.InfiniteResourceHandler;
+import net.neoforged.neoforge.transfer.handlers.templates.VoidResourceHandler;
+import net.neoforged.neoforge.transfer.resources.FluidResource;
+import net.neoforged.neoforge.transfer.resources.IResource;
+import net.neoforged.neoforge.transfer.resources.ItemResource;
+import net.neoforged.neoforge.transfer.transaction.TransactionManager;
+import org.assertj.core.api.Assertions;
+import org.assertj.core.api.ThrowableAssert;
+import org.junit.jupiter.api.MethodOrderer;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestMethodOrder;
+
+@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
+public class InstancedHandlerTests {
+    @Test
+    void emptyHandlers() {
+        //EMPTY no operation handlers but should throw on index calls
+        testEmptyHandler(EmptyResourceHandler.instance(), ItemResource.EMPTY);
+        testEmptyHandler(EmptyResourceHandler.instance(), FluidResource.EMPTY);
+    }
+
+    @Test
+    public void voidHandlers() {
+        //VoidResourceHandlers destroys resources but doesn't allow extraction
+        testVoidResource(VoidResourceHandler.ITEM, ItemResource.EMPTY);
+        testVoidResource(VoidResourceHandler.FLUID, FluidResource.EMPTY);
+    }
+
+    @Test
+    public void endlessHandlers() {
+        //InfiniteResourceHandlers creates infinite of a specified resource, but doesn't allow insertion
+        testEndlessResource(FluidResource.of(Fluids.WATER), FluidResource.EMPTY);
+        testEndlessResource(ItemResource.of(Blocks.COBBLESTONE), ItemResource.EMPTY);
+    }
+
+    private static <T extends IResource> void testVoidResource(VoidResourceHandler<T> handler, T emptyResource) {
+        Assertions.assertThat(handler.size()).withFailMessage("Size should be 1").isEqualTo(1);
+        Assertions.assertThat(handler.supportsExtraction()).withFailMessage("Extraction should be not be allowed").isFalse().isEqualTo(handler.supportsExtraction(0));
+        Assertions.assertThat(handler.supportsInsertion()).withFailMessage("Insertion should be be allowed").isTrue().isEqualTo(handler.supportsInsertion(0));
+        Assertions.assertThat(ResourceHandlerUtil.isValid(handler, emptyResource)).withFailMessage("Voids should always be valid for a resource").isTrue();
```

Comment:
Because of the short-circuit in `ResourceHandlerUtil#isValid` for the empty resource, this doesn't actually validate that void is always valid for a resource